PF-session3/pf3_assignment1/Remainder.java

```
@@ -0,0 +1,33 @@
+
+/**
+ * This class has method to find the remainder when x is divided by y.
+ *
+ * @version 1 24 Jan 2018
+ * @author Akshita Dixit
+ */
+
+package pf3_assignment1;
+
+public class Remainder {
+
+	/**
+	 * this is a recursive method to find the remainder when x is divided by y.
+	 * @param x
+	 * @param y
+	 * @return
+	 */
+	int rem(int x, int y)
+	{
+		if( x >=0 && y > 0)                               //mandatory condition for dividing two nos.
+		{
+			if(x > y)
+				return rem(x - y, y);                     //recursive call by each time reducing x by y and keeping y as second param.
+			else if(x == y)                               //condition when x is completely divisible by y.
+				return 0;
+			else 
+				return x;                                 //when x<y then x is the remainder.
+		}
+		else
+			return x;                                    //if y=0;
+	}
```

- The else block will be executed not only for `y = 0`, but also when `x < 0`. 

- Since it is not given in the question, what to do when `y <= 0` or `x < 0`, instead of assuming that we can simply `return x`, throw an exception. The exception thrown depends on the context. Sometimes you have to use your own custom exception class, while at other times, you can use from one of the existing exceptions. For example, in this case, you can use [ArithmeticException](https://docs.oracle.com/javase/8/docs/api/java/lang/ArithmeticException.html).

- Also, validations should be the first thing in the method, when possible. This helps in readability, and also follows the "fail-fast" principle, which is a recommended practice to follow:
-- https://dzone.com/articles/fail-fast-principle-in-software-development
-- https://www.martinfowler.com/ieeeSoftware/failFast.pdf